pytorch_lightning/callbacks/intel_compressor.py

```diff
@@ -0,0 +1,116 @@
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+
+
+class QuantizationMode(Enum):
+    PTQ_DYNAMIC = "post_training_dynamic_quant"
+    PTQ_STATIC = "post_training_static_quant"
+    QAT = "quant_aware_training"
+
+
+class INCQuantization(Callback):
+    """Quantization allows speeding up inference and decreasing memory requirements by performing computations and
+    storing tensors at lower bitwidths (such as INT8) than floating point precision. And this callback will
+    quantized model with Intel Neural Compressor(INC) tool.
+
+    Args:
+
+        config_path_or_obj: config file or Quantization_Conf class for INC.
+
+        monitor: Specified metric which will be computed in evaluation function.
+
+        module_name_to_quant: the model name which you want to quantize, it should be torch.nn.Module.
+
+        dirpath: save path where the quantization config and weights be saved to.
+
+    .. Intel Neural Compressor: https://github.com/intel/neural-compressor
+    """
+
+    def __init__(
+        self,
+        config_path_or_obj,
+        monitor: str,
+        module_name_to_quant: str,
+        datamodule: LightningDataModule = None,
+        dirpath: str = None,
+    ) -> None:
+        try:
+            from neural_compressor.conf.config import Quantization_Conf
+        except ImportError as e:
+            raise RuntimeError("Please install neural-compressor with: pip install neural-compressor") from e
+
+        self.config = (
+            config_path_or_obj
+            if isinstance(config_path_or_obj, Quantization_Conf)
+            else Quantization_Conf(config_path_or_obj)
+        )
+        self.approach = self.config.usr_cfg.quantization.approach
+        assert monitor is not None, "Please set metric name for evaluation loop!"
+        self.monitor = monitor
+        assert (
+            module_name_to_quant is not None
+        ), "Please set module_name_to_quant with module name which you want to quantize"
+        self.module_name_to_quant = module_name_to_quant
+        self.datamodule = datamodule
+        self.dirpath = dirpath
+
+    def on_fit_start(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule") -> None:
+        """Called when the compressor loop begins."""
+        try:
+            from neural_compressor.experimental import Quantization
+        except ImportError as e:
+            raise RuntimeError("Please install neural-compressor with: pip install neural-compressor") from e
+
+        def eval_func(model):
+            setattr(pl_module, self.module_name_to_quant, model)
+            out = trainer.validate(pl_module, self.datamodule.val_dataloader() if self.datamodule is not None else None)
+            return out[0][self.monitor]
+
+        trainer.quantizing = True
+
+        if self.dirpath is None:
+            self.dirpath = trainer.default_root_dir
+
+        quantizer = Quantization(self.config)

```

Comment:
As shared above, I believe the Callback should be subclassing `Quantization` or even its base primitive, but not be instantiated internally. This is a hack and won't work well with other callbacks.

Reply:
Agreed. @PenghuiCheng

Reply:
I will pushing some changes here.